### DISCUSSION TITLE
Creating an WTForms Extension for Quart

### DISCUSSION THREAD

**Discussion by crood58** [pallets/quart#182] [Q&A]
I am working on create a Quart version of Flask WTF. I have been successful this far, but I am having a hard time getting the CSRF extension to work and one test for the form CSRF. 

The github repo can be found at https://github.com/Quart-Addons/quart-wtf. 

In terms of the form CSRF, only one tests fails. This test looks is in test_csrf_form.py and is the following:

```
@pytest.mark.asyncio
async def test_form_csrf(app, client):
    """
    Test form CSRF.
    """
    @app.route("/", methods=["GET", "POST"])
    async def index():
        form = await QuartForm.create_form()

        if await form.validate_on_submit():
            return "good"

        if form.errors:
            return form.csrf_token.errors[0]

        return form.csrf_token.current_token

    async with app.app_context():
        response = await client.get("/")
        assert await response.get_data(as_text=True) == g.csrf_token

        response = await client.post("/")
        assert  await response.get_data(as_text=True) == "The CSRF token is missing."

        response = await client.post("/", data={"csrf_token": g.csrf_token})
        assert await response.get_data(as_text=True) == "good"
```
This test is failing since the form is valid on submit and is returning that the CSRF token is missing instead of good. I am not sure why this is. Any suggestions?

The CSRF extension will only pass the render token test. All other tests fail and I am not sure why and I have tried and tried to get it to work. 

Just looking for another set of eyes other than my own. Any help would be appreciated. 

Thank you in advance, 

Chris

**Comment by jenish2014** [comment]
Hi, I was looking for From-Handling for Quart and stumbled upon this discussion post. Its been a long time since this was posted and I am surprised no one has replied yet. 
Anyways, take a look at [starlette-wtf](https://github.com/muicss/starlette-wtf) implementation, maybe it will help.

**Comment by crood58** [comment]
> Hi, I was looking for From-Handling for Quart and stumbled upon this discussion post. Its been a long time since this was posted and I am surprised no one has replied yet. 
> 
> Anyways, take a look at [starlette-wtf](https://github.com/muicss/starlette-wtf) implementation, maybe it will help.

Yes, I have seen this and what I am using as a reference to build Quart-WTF along with Flask WTF. Tried using Flask WTF for CSRF but having a hard time getting to work. Starlette WTF uses middleware for this.
